File: java/concurrency/lock/RwLock.java
package concurrency.lock;

/**
 * @Author john
 * @Description 读写锁
 * @Date 2022/9/28 下午10:06
 */
public interface RwLock {
    /**
     * 写锁
     */
    void lock();

    /**
     * 写解锁
     */
    void unlock();

    /**
     * 读锁
     */
    void rLock();

    /**
     * 读解锁
     */
    void rUnlock();
}

File: golang/concurrency/lock/rwlock/rwlock.go
package rwlock

import "sync"

type RWLock interface {
	Lock()
	UnLock()
	RLock()
	RUnLock()
}

// 两个锁实现读写锁
type MyRWLock struct {
	c    int
	r, g sync.Mutex
}

func (rw *MyRWLock) Lock() {
	rw.g.Lock()
}

func (rw *MyRWLock) UnLock() {
	rw.g.Unlock()
}

func (rw *MyRWLock) RLock() {
	rw.r.Lock()
	defer rw.r.Unlock()
	rw.c++
	if rw.c == 1 {
		rw.g.Lock()
	}
}

func (rw *MyRWLock) RUnLock() {
	rw.r.Lock()
	defer rw.r.Unlock()
	if rw.c == 0 {
		return
	}
	rw.c--
	if rw.c == 0 {
		rw.g.Unlock()
	}
}
